Experimental (#1)
mpi-advance/mpi-advance: README.md

# MPI Advance
This repository contains packages and publicly released libraries that comprise the MPI Advance Project. The MPI Advance Project currently consists of the following libraries:

- Partitioned Communication: implements basic partitioned communication for MPI 3 compatible implementations (See MPI 4.0 Standard). [MPIPCL](https://github.com/mpi-advance/MPIPCL)
- Locality_aware: Collective algorithms for topology aware and sparse matrix communication. [Locality_Aware](https://github.com/mpi-advance/locality_aware)
- Stream-triggering: scheduled triggering of communication function by GPUs. [Stream-Triggering](https://github.com/mpi-advance/stream-triggering

## Building
This repository can be used to build some or all the current releases of the MPIAdvance libraries. The libraries included can be built from source with cmake or installed using the supplied spack packages. 

### Building from source

> 

Reviewer: The first option to build the MPI Advance set of libraries is to build everything from this repository directly.

#### Prerequisites
- MPI implementation supporting MPI 3.0+
- CMake 3.17+
- C11+

#### Initialize & Update Submodules

> Initialize the submodules, by running `git submodule init` in the top level of the repository. Each submodule is set to pull the latest version of the associated library. After the submodules have been initialized, they can be updated to the latest version by running. `git submodule update --remote` 

Reviewer: ```suggestion
Initialize the submodules, by running `git submodule init` in the top level of the repository (if not already cloned). Each submodule is set to pull the latest version of the associated library. After the submodules have been initialized, they can be updated to the latest version by running. `git submodule update --remote` 
```

After the submodules are up to date run the following to install the libraries. This library has limited 
To build each library, navigate to the top level of each submodule. 
```
	mkdir build
	cd build
	cmake .. <CMAKE OPTIONS> 
	make install

> ```

Reviewer: ````suggestion
After checking that all the submodules are present, the build process follows a normal CMake build:
```
   <cd to top level of MPI Advance repository>
	mkdir build
	cd build
	cmake .. <CMAKE OPTIONS> 
	make install
```
````

##### CMAKE Options
- `-DMPIA_PC` (ON): Build the Partitioned Communication library
- `-DMPIA_ST` (ON) : Build the Stream-triggering Library

> - `-DMPIA_LA` (ON) : Build the Locality_Aware Library

Reviewer: ```suggestion
This repository offers the ability to choose which MPI Advance libraries get built. NOTE: this will affect the final CMake package file that gets installed. For example, if you choose to only build the MPIPCL library, then doing `find_package(MPIAdvance)` will only attempt to look for `MPIPCL`. 
- `-DMPIA_PC` (ON): Build the Partitioned Communication library
- `-DMPIA_ST` (ON) : Build the Stream-triggering Library
- `-DMPIA_LA` (ON) : Build the Locality_Aware Library

In addition to selection which libraries get build, there are other options that influence some of the libraries in the collection: 
```

- `-DHIP` (OFF) : Build libraries with HIP Support. Build the library with HIP support. In order to use this option you may need to provide the target GPU architecture if it is not set in the environment. The specific architecture of an AMD GPU can be found by searching for the GPU model at ROCM capability list and supplying the compute capability via -DCMAKE_HIP_ARCHITECTURES. For example: `-DCMAKE_HIP_ARCHITECTURES=gfx942`
- `-DCXI` (OFF) : Build stream-triggering Library with CXI backend
- `-DHPE` (OFF) : Build stream-triggering library with HPE backend
- `-DBUILD_TESTS` (OFF) : Build available examples and tests 

While these options comprise a majority of the available build options for each library, additional options may be available when building individual libraries, please check the official repositories if more control is necessary for the build. 

> 

Reviewer: ```suggestion
#### Lightweight Build
A variant of the CMake build that only installs the final CMake package files is also offered by this repository. This style of build may be helpful for when another project needs to use `find_package(MPIAdvance)`, but already has a collection of MPI Advance libraries installed. In this case, setting `-DBUILD_LIGHT=ON` (along with the correct packages to include), will prevent this repository from attempting to build and install any MPI Advance library, and instead only install the final CMake package file.
```

#### Add the package repo to Spack
Spack packages for each library are provided as a spack_package repo inside the named folder. 
To add the repo to your installation of spack.

1. Clone spack from https://github.com/spack/spack if not already installed 
1. Activate spack, `. spack/share/spack/setup-env.sh`
2. Navigate inside the spack_packages directory. 
3. Use `spack repo add mpiadvance`
4. Run spack repo list to verify that spack has found and linked the repo. 

#### Installing the Spack Packages
Inside the provided spack package repo, there is a individual package for each library well as a wrapper package that will install the others as dependencies.

**Varients:**
- ~PC: Do not build Partitioned Communication
- ~ST: Do not build Stream-triggering
- ~LA: Do not bUILD Locality_aware
- +CUDA: Build with CUDA support**
- +ROCM: Build with HIP support.** 
- test: Enable unit testing where available. 

**If installing with CUDA or HIP support, you must provide either the `cuda_arch` or `hip_architecture` in your Spack environment or via command-line. 

#### Using the Spack Packages
You will need to load the package to add it to your path. `spack load <package name>

> If the package is not found make sure it has been installed. `spack find`

Reviewer: ```suggestion
If the package is not found make sure it has been installed with `spack find`
```

> Once loaded the packages can be found by cmake using the find_package and linking the package to your target. All libraries are installed under the MPIAdvance namespace. 

Reviewer: ```suggestion
Once loaded the packages can be found by CMake using the find_package and linking the package to your target. All libraries are installed under the MPI Advance namespace. 
```
